### PR TITLE
[mlir][GPU] Improve handling of GPU bounds

### PR DESCRIPTION
**Pull request by krzysz00**
This change reworks how range information for GPU dispatch IDs (block IDs, thread IDs, and so on) is handled.

1. `known_block_size` and `known_grid_size` become inherent attributes of GPU functions. This makes them less clunky to work with. As a consequence, the `gpu.func` lowering patterns now only look at the inherent attributes when setting target-specific attributes on the `llvm.func` that they lower to.
2. At the same time, `gpu.known_block_size` and `gpu.known_grid_size` are made official dialect-level discardable attributes which can be placed on arbitrary functions. This allows for progressive lowerings (without this, a lowering for `gpu.thread_id` couldn't know about the bounds if it had already been moved from a `gpu.func` to an `llvm.func`) and allows for range information to be provided even when `gpu.*_{id,dim}` are being used outside of a `gpu.func` context.
3. All of these index operations have gained an optional `upper_bound` attribute, allowing for an alternate mode of operation where the bounds are specified locally and not inherited from the operation's context. These also allow handling of cases where the precise launch sizes aren't known, but can be bounded more precisely than the maximum of what any platform's API allows. (I'd like to thank @benvanik for pointing out that this could be useful.)

When inferring bounds (either for range inference or for setting `range` during lowering) these sources of information are consulted in order of specificity (`upper_bound` > inherent attribute > discardable attribute, except that dimension sizes check for `known_*_bounds` to see if they can be constant-folded before checking their `upper_bound`).

This patch also updates the documentation about the bounds and inference behavior to clarify what these attributes do when set and the consequences of setting them up incorrectly.

### CODE REVIEW

**Comment by joker-eph** [comment]
I'm missing on the motivation here? Shouldn't we just have all GPU code be inside `gpu.func` anyway?

**Comment by krzysz00** [comment]
1. I remember there being a note on Discourse recently saying that non-kernel GPU functions could be `func.func`s
2. IREE keeps its GPU code in a `func.func` and I want those `InferIntRangeInterface` implementations to work there

**Comment by krzysz00** [comment]
Oh, and, 3, this'll make it work for `spirv.func` if that's a thing

**Comment by joker-eph** [comment]
> I remember there being a note on Discourse recently saying that non-kernel GPU functions could be func.funcs

Yes, but as far as I remember, the only reason for this is because we wanted to allow function that works on CPU and GPU: that is if a function does not make use of GPU specific features then it can be a func.func.

> Oh, and, 3, this'll make it work for spirv.func if that's a thing

But wouldn't a spirv.func be the result of this conversion? That is you perform dialect conversion from the GPU dialect to SPIRV and you get a spirv.func with the spirv operation matching this GPU operation?

File: mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp [hunk 1]
```diff
@@ -73,12 +85,16 @@ static std::optional<uint64_t> getKnownLaunchDim(Op op, LaunchDims type) {
...
     case LaunchDims::Block:
-      return llvm::transformOptional(func.getKnownBlockSize(dim), zext);
+      return llvm::transformOptional(
+          getKnownLaunchAttr(func, GPUFuncOp::getKnownBlockSizeAttrName(), dim),
+          zext);
     case LaunchDims::Grid:
-      return llvm::transformOptional(func.getKnownGridSize(dim), zext);
+      return llvm::transformOptional(
+          getKnownLaunchAttr(func, GPUFuncOp::getKnownGridSizeAttrName(), dim),
+          zext);
```
Comment (joker-eph) [inline mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: This seems iffy: it dispatches to discardable attributes instead of inherent attrs/properties.

**Comment by krzysz00** [comment]
During the conversion, the surrounding context could either be the source function (say, `gpu.func`) or the target function (say `llvm.func`) depending on if the target code gets the surrounding function or its body gets rewritten first, hence the original conversion code having cases for both GPU and LLVM's func ops, which this change generalizes

The known sizes also aren't inherent attributes of `gpu.func` either - see https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/GPU/IR/GPUOps.td#L434 , which goes to a discardable tag

File: mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp (hunk 1 above)
Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: I checked the implementation, and indeed the `getKnownBlockSize()` is just a wrapper around the discardable attributes: but that seems like a bug to me? It should be inherent (and optional) I think.

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: It's house style for `gpu.func` (see also the `kernel` attribute, last I checked)

And I found my cute for "GPU code doesn't have to be in a GPU func" https://discourse.llvm.org/t/mlir-gpu-calling-gpu-func-that-is-not-a-kernel-a-k-a-device-function/79285 .

And so there could be a pass that propagates these annotations from kernels to the `func.func`s they call, which is an argument against making them inherent properties of GPU functions only.

So I'd argue that the known block size annotations are dialect-level metadata and that the fix here it to move the name to GPUDialect (and probably to go to a new-style attribute helper)

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: > And so there could be a pass that propagates these annotations from kernels to the func.funcs they call, which is an argument against making them inherent properties of GPU functions only.

I would rather see a process where such func.func get turned into proper gpu.func instead.
Discardable attributes are supposed to be ... discardable :)

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: Maybe there should be a `GPUFuncOpInterface`, I don't know, that seems heavy and it's not clear we should add func.func to it either, so it does not really help.

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: I think that functions have, in practice, stronger semantics re discardability, similarly to how modules do. That is, while `dlti.dl_spec` is "discardable" (even on `builtin.module`), it's expected that passes will preserve attributes like those unless they're something like an inliner that you know won't handle that universe of discardable information.

But that overall point aside ... `gpu.known_block_size` is discardable. It's an optimization hint. Dropping it halfway down your pipeline could cause optimization issues but shouldn't cause miscompilation.

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: > I think that functions have, in practice, stronger semantics re discardability, similarly to how modules do.

That's a possible view, I'm on the side that we should define proper inherent attributes for what we intend to support instead (including DLTI for example).

Anyway: stepping back we still need to get back at the intended flow of GPU compilation: in particular whether GPU-specific features should be tied to specific containers or not.

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: I've found at least one downstream context (IREE, where I want to use `-int-range-optimizations`) that has decided the answer is "no" (and even in rocMLIR, we keep our GPU-specific code in a `func.func` until relatively late in our pipeline).

To try and cook up an argument for no, I'd say that GPU-specific code can arise from some notional `-specialize-to-target` pass that takes a `func.func` (or some other `*.func`) and makes clones of it for different targets. Forcing such a pass to change the function operation (and, worse, the nesting structure, because a gpu.func must be in a gpu.module, which is isolated from above, so you'd have to copy any referenced globals in etc) just to get access to `gpu.thread_id` seems to be to be a rather strange decision.

That is, the state of the world is "upstream MLIR provides a GPU compilation infrastructure that's intended to be used as a whole, including gpu.module/gpu.func. However, the wrappers around GPU primitives (like gpu.thread_id or the matrix ops) can be used independently of that infrastructure"

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: > I've found at least one downstream context (IREE, where I want to use -int-range-optimizations) that has decided the answer is "no"

Sure, but that's not really helping me reasoning here: I see this as completely anecdotical actually.

> To try and cook up an argument for no, I'd say that GPU-specific code can arise from some notional -specialize-to-target pass that takes a func.func (or some other *.func) and makes clones of it for different targets. Forcing such a pass to change the function operation (and, worse, the nesting structure, because a gpu.func must be in a gpu.module, which is isolated from above, so you'd have to copy any referenced globals in etc)) just to get access to gpu.thread_id seems to be to be a rather strange decision.

Why strange? That seems to be what I expect from anyone targeting the GPU dialect actually?

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: One other note is that the GPU dialect already has two official, somewhat independent purposes:
1. Handling compilation of GPU device code (`gpu.module`, `gpu.func`, etc.)
2. Handling execution of GPU code (`gpu.launch_func` etc.)

and I'm pretty sure you're allowed to just use one half and replace the other half with your own custom arrangement for that same task and have everything basically work (if slightly less seamlessly).

My claim is that operations like `gpu.thread_id` and `gpu.shuffle` are in a third class: abstractions around platform-specific GPU intrinsics. That is, these are operations meant to abstract across what are almost inevitably platform-specific intrinsics, allowing people to write code generation schemes that can target "a GPU" (though somewhere in their context they're likely to know which).

The abstracting segment of the GPU dialect is like, say, `math.sin` - there isn't and, IMO, shouldn't, be any special workflow needed to use it.

--- 

And re how GPU compilation happens, consider the following hypothetical compiler
```mlir
something.multi_target_function [..., @f] targets [
  builtin.module {something.target_info = <... cpu ... >} {
    func.func @f(...) { ... }
  },
  builtin.module {sometihng.target_info = <... gpu ...>} {
    func.func @f(...) { ... }
  }
]
```

If we forced all uses of basic GPU intrinsics into a `gpu.module`, then you'd have a hard time defining
```
std::optional<something::TargetInfoAttr> getTargetInfo(mlir::FunctionOpInterface func);
```
in a way that doesn't know about the special-case that is a GPU module. And attaching that data to the `gpu.module` instead means that anyone walking over `targets` has to watch out for `gpu.container_module` and recurse into the builtin.

I don't think that this extra level of nesting is justified to expose the GPU dialect's wrapper operations. Downstream developers should be able to use their own per-target isolation/separation scheme without losing access to the useful lowerings  that live in the GPU dialect.

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: Or, to put it another way that just came to me, nothing about the GPU dialect's portability wrappers requires the use of the gpu.module/gpu.func infrastructure and there's no good reason I can think of to couple them. (and `gpu.thread_id x` is one such portability wrapper).

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: > One other note is that the GPU dialect already has two official, somewhat independent purposes:

Yes! This is something we should document better, it's "almost" two separate dialects (let's say "sub-dialects" ;))

> and I'm pretty sure you're allowed to just use one half and replace the other half with your own custom arrangement for that same task and have everything basically work (if slightly less seamlessly).

That seems like a bug to me :)
If we want the launch_func to accept other implementations for a kernel, then I would argue we should have a KernelEntryPointOpInterface (right now the verifier just checks for `FunctionOpInterface` which does not seem like the right check for running a kernel IMO). Anyway it does not invalidate your point.

> My claim is that operations like gpu.thread_id and gpu.shuffle are in a third class: abstractions around platform-specific GPU intrinsics. That is, these are operations meant to abstract across what are almost inevitably platform-specific intrinsics, allowing people to write code generation schemes that can target "a GPU" (though somewhere in their context they're likely to know which).

That seems fair, thanks for elaborating on this.
Can we start by working on a documentation update for the dialect and describe these 3 categories in the dialect?
1) Host code manipulation and offloading support
2) "Container" object (gpu.module, gpu.func)
3) GPU specific operations (to be mixed with arith and others).


> If we forced all uses of basic GPU intrinsics into a gpu.module, then you'd have a hard time defining
> std::optional<something::TargetInfoAttr> getTargetInfo(mlir::FunctionOpInterface func);

I don't follow this part, I am concerned that you're saying that we **can't** use a gpu.module, when this is where the GPU target info should live?

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: Re `getTargetInfo`, my claim is that a user of the GPU dialect should not be (and, by current practice, is not) required to use a `gpu.module` as the container for their GPU compilations. They should be able to use a `their_own_custom.module` that's annotated with "this is targeting a GPU". (This gets even more relevant when we start merging the target attribute RFCs - I should be able to use a `builtin.module attributes {target = #gpu.rocdl_target<...>]` just as much as a `gpu.module`.)

The problem with forcing a gpu.module, especially since it requires its immediate parent to have `gpu.container_module` on it, is that it breaks a general multi-target compilation scheme. GPU modules would have to have an extra level of nesting when there isn't a corresponding, say, `x86.module`.

(Specifically re the attributes we're discussing, I think that `gpu.known_block_size` is a general piece of infrastructure that can be used in flows other than the gpu.module/gpu.func one. The contract with those is that "the function that contains the block_id/thread_id/... operations should have these attributes set if they're to be used in optimization or lowering". That can be any kind of function. Perhaps `gpu.func` could have special inherent attributes to make it harder to lose the hint, but ... consider that you can lower `gpu.func` to `llvm.func` or `spirv`func` in a separate sequence of conversions from the lowering of `gpu.thread_id x` ... and just copy the `gpu.known_block_sizes` attribute over. Heck, the existing gpu-to-{nvvm,rocdl} passes do this and then delete the attribute at the end)

In summary, insisting on gpu.module for GPU functions breaks other possible abstractions for keeping the code for different devices separate.

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: That being said, you've got a point about the problems with discardable attributes, so how about ... we do both?

Known block/grid sizes would be an optional inherent attribute of `gpu.func`, and attempts to retrieve that data would go looking for those values if available. However, when the parent of the `gpu.*_id` or `gpu.*_dim` is some other kind of function, the data in question would live in `gpu.known_*_sizes`, a discardable attribute.

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: > Re getTargetInfo, my claim is that a user of the GPU dialect should not be (and, by current practice, is not) required to use a gpu.module as the container for their GPU compilations. They should be able to use a their_own_custom.module that's annotated with "this is targeting a GPU". 

Sure, but you're leaving out some parts here about "how much of upstream can they expect to rely on when they are in this situation".  The rubber hits the road when you try to write **all** code without being able to assume anything about the structure or the operations used: having things compose is nice, but that does not mean that the upstream lowering would just work out of the box regardless.
(otherwise why even both introducing things like gpu.func and gpu.module at all?)

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: > The problem with forcing a gpu.module, especially since it requires its immediate parent to have gpu.container_module on it, is that it breaks a general multi-target compilation scheme. GPU modules would have to have an extra level of nesting when there isn't a corresponding, say, x86.module.

I didn't follow this part?

> In summary, insisting on gpu.module for GPU functions breaks other possible abstractions for keeping the code for different devices separate.

Even if this diagnostic is true (I haven't followed why?), it's not obvious to me that there is a single answer: that is can we fix whatever limitation you feel exist around gpu.module to make it usable in such situation?

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: Yeah, honestly, I don't think gpu.module and gpu.func should exist.

I think `gpu.module` should be `builtin.module {target = #gpu.{nvvm,rocdl,...}_target<...>` and that `gpu.func` should be `func.func {gpu.kernel}`.

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: And, on this specific question, the GPU ID operations already keyed off a discardable optimization hint that that could be on a `llvm.func` as well, so I figure we might as well make the generalization we already had two special cases of anyway.

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: (The counterpoint there is that there's no currently good way to run a pass pipeline on "all modules that have an attribute", so that's why we don't do it this way, but still)

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: > Yeah, honestly, I don't think gpu.module and gpu.func should exist.
> I think gpu.module should be builtin.module {target = #gpu.{nvvm,rocdl,...}_target<...> and that gpu.func should be func.func {gpu.kernel}.

That's a possibility, but I'm concerned about this being overly lose :(

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: My thoughts on that target thing would be to implement those same interfaces for `rocdl.target` and the like.

And ... having looked around some more, it seems like `gpu.module` doesn't have the requirement that it live inside a `gpu.container_module` anymore (unless you want to use `gpu.launch_func`).

Comment (krzysz00) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: Though I think we just drifted from functions to modules.

As far as I'm concerned, the only difference between

```
gpu.func @f(...) known_block_sizes = [64, 2, 1] {
  ...
}
```

and
```
func.func @f(...) attributes {gpu.known_block_sizes = [64, 2, 1]} {
  ...
}
```
(or even `my_custom.func` in the above) is the somewhat higher risk of discardability on the optimization hint.

This PR's trying to generalize some already general handling for this particular hint so that we're not locking downstream into needing a very particular infrastructure (that might not compose with other compilation infrastructures because there's an extra level of module involved) just to get `ThreadIdOp::inferIntRanges()` working.

File: mlir/lib/Conversion/GPUCommon/IndexIntrinsicsOpLowering.h [hunk 2]
```diff
@@ -57,11 +57,7 @@ struct GPUIndexIntrinsicOpLowering : public ConvertOpToLLVMPattern<Op> {
       break;
     }
 
-    Operation *function;
-    if (auto gpuFunc = op->template getParentOfType<gpu::GPUFuncOp>())
-      function = gpuFunc;
-    if (auto llvmFunc = op->template getParentOfType<LLVM::LLVMFuncOp>())
-      function = llvmFunc;
+    Operation *function = op->template getParentOfType<FunctionOpInterface>();
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUCommon/IndexIntrinsicsOpLowering.h:60]: Generalizing this seems fine to me until we have a non-discardable attribute on `gpu.func`

File: mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp (hunk 1 above)
Comment (kuhar) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: > My claim is that operations like gpu.thread_id and gpu.shuffle are in a third class: abstractions around platform-specific GPU intrinsics. That is, these are operations meant to abstract across what are almost inevitably platform-specific intrinsics, allowing people to write code generation schemes that can target "a GPU" (though somewhere in their context they're likely to know which).

+1, this has always been my understanding and I'm not aware of any intended limitations as to where these have to live `gpu.func`/`func.func`/`spirv.func`/etc. None seem to be documented at the moment either AFAICT.

**Comment by kuhar** [comment]
I'd be leaning towards making this an inherent attribute over `gpu.func` but also allow this as a discardable attribute over other function ops. This is supported by the existing support over `llvm.func`. I don't have any strong options wrt to the sequencing here.

**Comment by joker-eph** [comment]
The API to access discardable and non-discardable API are supposed to be completely segregated though: hence why I mentioned the need for an interface.

File: mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp (hunk 1 above)
Comment (joker-eph) [reply mlir/lib/Dialect/GPU/IR/InferIntRangeInterfaceImpls.cpp:97]: >  None seem to be documented at the moment either AFAICT.

Right, hence why I [commented above](https://github.com/llvm/llvm-project/pull/95166#discussion_r1635557491) about the need to start with documenting our intent here. If we agree on the destination, it'll be easier to make fast progress on each step!

File: mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp [hunk 3]
```diff
@@ -74,35 +74,41 @@ GPUFuncOpLowering::matchAndRewrite(gpu::GPUFuncOp gpuFuncOp, OpAdaptor adaptor,
...
+  if (knownBlockSize)
+    attributes.emplace_back(
+        rewriter.getStringAttr(
+            gpu::GPUDialect::KnownBlockSizeAttrHelper::getNameStr()),
+        knownBlockSize);
+  if (knownGridSize)
+    attributes.emplace_back(
+        rewriter.getStringAttr(
+            gpu::GPUDialect::KnownGridSizeAttrHelper::getNameStr()),
+        knownGridSize);
```
Comment (joker-eph) [inline mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:101]:
```suggestion
  auto *gpuDialect = cast<GPUDialect>(gpuFuncOp.getDialect());
  if (knownBlockSize)
    attributes.emplace_back(gpuDialect.getKnownBlockSizeAttrHelper().getName()),
        knownBlockSize);
  if (knownGridSize)
    attributes.emplace_back(gpuDialect.getKnownGridSizeAttrHelper().getName()),
        knownGridSize);
```

The point of the ODS discardableAttrs support is (also) about avoiding to recreate StringAttr constantly :)

Comment (krzysz00) [reply mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:101]: ... Huh, I knew I was missing some obvious way to do that, thanks!

**Comment by krzysz00** [comment]
PR description has been updated to reflect the radical change in approach that came out of discussion

File: mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp [hunk 4]
```diff
@@ -349,27 +348,32 @@ static void populateOpPatterns(LLVMTypeConverter &converter,
...
+      converter, IndexKind::Grid, IntrType::Dim);
+  patterns.add<GPUReturnOpLowering>(converter);
   patterns.add<GPUFuncOpLowering>(
       converter,
       /*allocaAddrSpace=*/ROCDL::ROCDLDialect::kPrivateMemoryAddressSpace,
       /*workgroupAddrSpace=*/ROCDL::ROCDLDialect::kSharedMemoryAddressSpace,
-      ROCDL::ROCDLDialect::KernelAttrHelper(&converter.getContext()).getName());
+      ROCDL::ROCDLDialect::KernelAttrHelper(&converter.getContext()).getName(),
+      ROCDL::ROCDLDialect::ReqdWorkGroupSizeAttrHelper(&converter.getContext())
+          .getName());
```
Comment (joker-eph) [inline mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp:376]: Can you get a handle to the dialect first and then use it to get the various names?
```
      auto *rocdlDialect = converter.getContext().getLoadedDialect<ROCDLDiale```

**Review by joker-eph — APPROVED** [review]
That seems OK, but can we have a PR to document the GPU dialect layering? It would be great to not lose all the discussion we had here, it can be handy to have this encoded for future reference.

**Comment by krzysz00** [comment]
@joker-eph The documentation updates are at #95812

**Review by kuhar — APPROVED** [review]

**Comment by grypp** [comment]
I apologize for joining the discussion a bit late. I understand the rationale behind introducing the `known_block_size` attribute.

However, I have some concerns regarding the following code snippet (it's in the example). As per CUDA (or other gpu languages), these attributes are restricted to **GPU kernels**. It appears, based on my understanding, that the proposed changes in the PR are extending these attributes to `func.func`. 

Could you please clarify the intended behavior of the following code? Specifically, should it be treated as a GPU kernel, a GPU function, or a host function? 
```
  func.func @annotated_kernel()
    attributes {gpu.known_block_size = array<i32: 8, 12, 16>,
        gpu.known_grid_size = array<i32: 20, 24, 28>} {
```

If the goal is to permit GPU kernel attributes on `func.func`, in that case, this PR needs a verifier that allows `known_block_size/known_grid_size` only on gpu kernels.

**Comment by krzysz00** [comment]
@grypp The intent of that code is that the `func.func` *could* be modelling a GPU kernel when you're using a non-upstream lowering flow

**Comment by krzysz00** [comment]
That is, the `func.func` is either a host or device function depending on the pipeline's context / if it's in a GPU module / ...

**Comment by grypp** [comment]
> That is, the `func.func` is either a host or device function depending on the pipeline's context / if it's in a GPU module / ...

I understand your intention. However, I'm asking if we can use a verifier to ensure that the blocksize attribute cannot be present unless `func.func` is within the `gpu.module`?

**Comment by krzysz00** [comment]
> I understand your intention. However, I'm asking if we can use a verifier to ensure that the blocksize attribute cannot be present unless `func.func` is within the `gpu.module`?

The specific intent of this change is that that is *not* a requirement.

You are allowed to do 
```
some.other.thing target(SomeGPUTargetAttr) {
  func.func {gpu.known_block_size = array:i32: 64, 2, 1>} {
    ...
   %N = gpu.thread_id x
   ...
  }
}
```

**Comment by krzysz00** [comment]
If you set this attribute on a host function, the lowering for host functions will ignore it. If you had a `gpu.thread_id x` in a function that ends up being a host function ... lowering failure, since thhere's no XToLLVM for x86_64
